src/mscorlib/shared/System/ReadOnlySpan.cs

```diff
@@ -198,8 +198,18 @@ public bool IsEmpty
         /// </summary>
         public void CopyTo(Span<T> destination)
         {
-            if (!TryCopyTo(destination))
+            // Using "if (!TryCopyTo(...))" results in two branches: one for the length
+            // check, and one for the result of TryCopyTo. Since these checks are equivalent,
+            // we can optimize by performing the check once ourselves then calling Memmove directly.
```

Comment:
Semi-related to this PR, since you've been looking at the perf a lot, I've had a bunch of cases where I need to use {Try}CopyTo and I know from previous checks that the destination is long enough. Given that, I generally preferred to use TryCopyTo and then either ignore the return value or assert that it's true, but that was when CopyTo called TryCopyTo and I could avoid the same extra branch you mention in the comment. Thoughts on the right approach here? Any sense of how much the length check costs, and whether there's any benefit in exposing an "unsafe" variant that skips the check entirely?

Reply:
The length checks are cheap. I do not think we want to be avoiding them, even for corelib.

If there are situations where they are obviously redundant, we should make sure that the JIT is optimizing the redundancy out.

Reply:
Opened https://github.com/dotnet/coreclr/issues/16042 to track removing redundant comparisons.